Initial work to add `Request Transfer` buttons to the UI

daledupreez opened the pull request:
#### Changes proposed in this Pull Request

* Add a `Transfer Domain to WordPress.com` action for domains that are mapped in from elsewhere.

#### Testing instructions

* Verify that when you view the details for a mapped domain, a "Transfer Domain to WordPress.com" item appears in the nav options at the bottom of the screen.
  * When you click on that item, you should shift to the Transfer Domain UI with the domain entered in the search/input box.
  * When you click the "Back" link in the header, ensure that you go back to the domain management screen for the mapped domain.
* Verify that the link _does not_ appear for a mapped subdomain.
* Verify that when you try to add a domain, and then work through the "Use a domain I own" -> "Transfer to WordPress.com" flow, and then click "Back" in the header, you should be returned to the domain search screen.

### Open Issues/Questions

* Please take a **very close look** at the way I've handled the "Back" button logic, which involves a new URL parameter _and_ a call to `page.back()`. The original logic was very stateful, and I've added a less stateful approach, but it doesn't change the original logic (which dates back to the original implementation for domain transfers).

**NOTE:** This depends on back end change D33690-code.

a8ci18n commented on client/my-sites/domains/domain-management/edit/mapped-domain.jsx:
```diff
@@ -100,12 +105,20 @@ const MappedDomain = createReactClass( {
+						/>
+
+						<Button href={ getDomainTransferPath( selectedSite.slug, domain.name ) }>
+							{ this.props.translate( 'Request Transfer', {
```
Hi! I've found a possible matching string that has already been translated **27** times:
`translate( 'Request Transfer Code' )` *ES Score: 7*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

daledupreez replied:
👎

hambai commented:
1. I would put the button as navigation link (after "DNS records"). Maybe label it "Transfer your domain to WordPress.com" . It just seems that "Make Primary" and "Payment Settings" should always be there no matter if you're looking at a mapping or a domain registration.
2. We should not show this link if we don't support incoming transfers for this TLD. Also not show the link if this is a mapped subdomain (or maybe not?). Not sure if we can reliably check if the domain has expired - if we can and if it doesn't have significant impact on some API we can do that too. We also should not show the transfer link for .blog subdomains and for wpcomstaging.com.
3. We'll have to make the back button work as expected, but I think it should be relatively easy

klimeryk commented on client/my-sites/domains/paths.js:
```diff
@@ -121,10 +121,29 @@ export function domainMapping( siteName, domain = '' ) {
+	}
+
+	if ( params.length > 0 ) {
+		path += '?' + params.join( '&' );
```
At this point, I'd recommend using [`qs`](https://github.com/ljharb/qs) - we already use it in Calypso, so it's not a new dependency.

Example usage would be:

```
const query = {};
if ( domain ) {
    params.initialQuery = domain;
}

if ( useStandardBack ) {
    query.useStandardBack = true; // or 1
}

const queryString = qs.stringify( query );
if ( queryString ) {
    path += '?' + queryString;
}
```

Not exactly much shorter, but definitely less hacky and we get some security benefits as well, I'd imagine.

klimeryk commented on client/my-sites/domains/domain-management/edit/mapped-domain.jsx:
```diff
@@ -152,6 +157,22 @@ const MappedDomain = createReactClass( {
 	},
+
+	transferMappedDomainNavItem() {
+		const domain = this.props.domain;
```
This is definitely a nitpick, but in case it's helpful/interesting: the usual pattern is using a [destructuring assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment). It's slightly shorter, but also allows pulling in more stuff from `props` (which is usually the case). And it's more "future-proof" in case future PRs need additional stuff from props.

So, for example:

```
const { domain, translate } = this.props; // now you can use `translate` below, instead of `this.props.translate`
```

Up to you what props need to be shortened like this - the usual rule of thumb is if it's used more than once, do it. And if you're already destructuring props, pull in `translate` as well, even if it's used once.

klimeryk commented on client/my-sites/domains/transfer-domain/index.jsx:
```diff
@@ -30,6 +30,7 @@ import TrademarkClaimsNotice from 'components/domains/trademark-claims-notice';
 export class TransferDomain extends Component {
 	static propTypes = {
 		initialQuery: PropTypes.string,
+		useStandardBack: PropTypes.string,
```
This should be a boolean and the conversion from string to boolean should be done at the controller level.

klimeryk commented:
PS: also probably will need update after the backend field's name changes.

daledupreez commented:
@hambai, @klimeryk or @eltongo, please take a final look at this PR when you get a chance -- I updated from `master` and renamed one of the properties (tomorrow should be fine, BTW). I hit what looks like an unrelated failure in the e2e tests, so I am re-running them to be safe.

eltongo commented:
Tested and works as expected :)
